networkx/algorithms/approximation/steinertree.py

```diff
@@ -88,11 +49,20 @@ def steiner_tree(G, terminal_nodes, weight="weight"):
     .. [1] Steiner_tree_problem on Wikipedia.
        https://en.wikipedia.org/wiki/Steiner_tree_problem
     """
-    # H is the subgraph induced by terminal_nodes in the metric closure M of G.
-    M = metric_closure(G, weight=weight)
-    H = M.subgraph(terminal_nodes)
+    # M is the subgraph induced by terminal_nodes with shortest corresponding pairwise paths in G.
+
+    M = nx.Graph()
+    for u_index in range(len(terminal_nodes)):
+        u = terminal_nodes[u_index]
+
+        for v_index in range(u_index + 1, len(terminal_nodes)):
+            v = terminal_nodes[v_index]
+            dist_u_v = nx.shortest_path_length(G=G, source=u, target=v, weight=weight)
+            path_u_v = nx.shortest_path(G=G, source=u, target=v, weight=weight)
```

Comment:
I will leave it to the maintainers for a full review, but just to note that here there are two independent calculations of shortest path. It might be better to calculate only `path_u_v` and use that to construct `dist_u_v`, avoiding a potentially expensive re-calculation.